### PR TITLE
cifar10 resnet v1 and v2

### PR DESCRIPTION
**Pull request by roatienza**
Support for both ResNet v1 and v2
Added performance benchmarks
Use of learning rate scheduler instead of lr reducer to improve accuracy
Implementation as close as possible to the orig papers

### CODE REVIEW

**Comment by ahundt** [comment]
fyi https://github.com/farizrahman4u/keras-contrib/blob/master/keras_contrib/applications/resnet.py

**Comment by roatienza** [comment]
@ahundt - PR is for fixes in the master keras examples/cifar10_resnet.py

**Comment by roatienza** [comment]
In addition, the code in keras-contrib is derived from https://github.com/raghakot/keras-resnet which can only achieve 86% accuracy for ResNet18. In this PR, the same network can achieve ~92% accuracy.

**Comment by ahundt** [comment]
Yes, I was just pointing out that it exists. If you're aware of the source of that difference in performance would you mind submitting a patch to the keras-contrib version too?

**Comment by roatienza** [comment]
@ahundt - I will find why the keras-contrib version has a lower val_acc and submit a PR.

**Comment by ahundt** [comment]
thanks! I appreciate it!

File: examples/cifar10_resnet.py [hunk 1]
```diff
@@ -130,26 +164,31 @@
 # Instantiate and compile model.
 model = Model(inputs=inputs, outputs=outputs)
 model.compile(loss='categorical_crossentropy',
-              optimizer=Adam(),
+              optimizer=Adam(lr=lr_schedule(0)),
               metrics=['accuracy'])
 model.summary()
 
+if version == 2:
+    print("ResNet v2")
```
Comment (fchollet) [inline examples/cifar10_resnet.py]: Nit: use `'` for consistency with rest of file

File: examples/cifar10_resnet.py [hunk 2]
```diff
@@ -23,15 +25,46 @@
...
-use_max_pool = False
+num_filters = 16
+num_blocks = 3
+num_sub_blocks = 2 * n
+
+
+# Learning rate scheduler - called every epoch as part of callbacks
+def lr_schedule(epoch):
+    lr = 1e-3
+    if n == 18:
```
Comment (fchollet) [inline examples/cifar10_resnet.py]: Avoid the use of out of scope variables if they aren't all-caps globals.

File: examples/cifar10_resnet.py [hunk 3]
```diff
@@ -23,15 +25,46 @@
...
-num_blocks = 4
-num_sub_blocks = 2
-use_max_pool = False
+num_filters = 16
+num_blocks = 3
+num_sub_blocks = 2 * n
+
+
+# Learning rate scheduler - called every epoch as part of callbacks
+def lr_schedule(epoch):
```
Comment (fchollet) [inline examples/cifar10_resnet.py]: Such methods should have a docstring to explain what they do.

File: examples/cifar10_resnet.py [hunk 4]
```diff
@@ -60,6 +93,12 @@
 # Normalize data.
 x_train = x_train.astype('float32') / 255
 x_test = x_test.astype('float32') / 255
+
+if use_pix_mean:
```
Comment (fchollet) [inline examples/cifar10_resnet.py]: "subtract_mean_pixel" would be more explicit

File: examples/cifar10_resnet.py [hunk 5]
```diff
@@ -116,7 +149,8 @@
                        kernel_initializer='he_normal',
                        kernel_regularizer=l2(1e-4))(x)
         x = keras.layers.add([x, y])
-        x = Activation('relu')(x)
+        if version != 2:
+            x = Activation('relu')(x)
```
Comment (fchollet) [inline examples/cifar10_resnet.py]: Is the placement of the relu the only difference between v1 and v2?

File: examples/cifar10_resnet.py (hunk 3 above)
Comment (roatienza) [reply examples/cifar10_resnet.py]: Docstring added.

File: examples/cifar10_resnet.py (hunk 2 above)
Comment (roatienza) [reply examples/cifar10_resnet.py]: No need for special case of n == 18 in the latest commit.

File: examples/cifar10_resnet.py (hunk 4 above)
Comment (roatienza) [reply examples/cifar10_resnet.py]: subtract_pixel_mean variable name is used instead of use_pix_mean.

File: examples/cifar10_resnet.py (hunk 5 above)
Comment (roatienza) [reply examples/cifar10_resnet.py]: I reviewed the 2 papers again. Resnet v1 is accurate. I may have overlooked some details in Resnet v2. Thus, I revised Resnet v2 following the Torch implementation by K He. Basically, Resnet v1 is characterized by post activation ( 2 3x3 Conv-BN-ReLU) while Resnet v2 is pre-activation (1x1 BN-ReLU-Conv, 3x3 BN-ReLU-Conv, 1x1 BN-ReLU-Conv or known as bottleneck layer).  Resnet v1 number of parameters are approx the same in paper of K He.

File: examples/cifar10_resnet.py (hunk 1 above)
Comment (roatienza) [reply examples/cifar10_resnet.py]: Revised using ''.

**Comment by ahundt** [comment]
This is starting to look like it duplicates the resnet model functions in keras_contrib, perhaps it would make sense to move that into keras/applications?

**Comment by fchollet** [comment]
That would be fine to add in `applications/resnet.py`. We already have `ResNet50`, so surely the amount of changes would be limited?

**Comment by fchollet** [comment]
Note that if we add it in `applications`, each added model should have a pretrained imagenet weights file.

**Comment by ahundt** [comment]
@fchollet [https://github.com/farizrahman4u/keras-contrib/blob/master/keras_contrib/applications/resnet.py](keras_contrib/applications/resnet.py) is much more parameterized to support different problem types than ResNet50, so it is quite different, but I think these are problems that could be remapped similarly to other past API updates. Here are the details:

```python

def ResNet(input_shape=None, classes=10, block='bottleneck', residual_unit='v2', repetitions=None,
           initial_filters=64, activation='softmax', include_top=True, input_tensor=None, dropout=None,
           transition_dilation_rate=(1, 1), initial_strides=(2, 2), initial_kernel_size=(7, 7),
           initial_pooling='max', final_pooling=None, top='classification'):
    """Builds a custom ResNet like architecture. Defaults to ResNet50 v2.
    Args:
        input_shape: optional shape tuple, only to be specified
            if `include_top` is False (otherwise the input shape
            has to be `(224, 224, 3)` (with `channels_last` dim ordering)
            or `(3, 224, 224)` (with `channels_first` dim ordering).
            It should have exactly 3 inputs channels,
            and width and height should be no smaller than 8.
            E.g. `(224, 224, 3)` would be one valid value.
        classes: The number of outputs at final softmax layer
        block: The block function to use. This is either `'basic'` or `'bottleneck'`.
            The original paper used `basic` for layers < 50.
        repetitions: Number of repetitions of various block units.
            At each block unit, the number of filters are doubled and the input size is halved.
            Default of None implies the ResNet50v2 values of [3, 4, 6, 3].
        transition_dilation_rate: Used for pixel-wise prediction tasks such as image segmentation.
        residual_unit: the basic residual unit, 'v1' for conv bn relu, 'v2' for bn relu conv.
            See [Identity Mappings in Deep Residual Networks](https://arxiv.org/abs/1603.05027)
            for details.
        dropout: None for no dropout, otherwise rate of dropout from 0 to 1.
            Based on [Wide Residual Networks.(https://arxiv.org/pdf/1605.07146) paper.
        transition_dilation_rate: Dilation rate for transition layers. For semantic
            segmentation of images use a dilation rate of (2, 2).
        initial_strides: Stride of the very first residual unit and MaxPooling2D call,
            with default (2, 2), set to (1, 1) for small images like cifar.
        initial_kernel_size: kernel size of the very first convolution, (7, 7) for imagenet
            and (3, 3) for small image datasets like tiny imagenet and cifar.
            See [ResNeXt](https://arxiv.org/abs/1611.05431) paper for details.
        initial_pooling: Determine if there will be an initial pooling layer,
            'max' for imagenet and None for small image datasets.
            See [ResNeXt](https://arxiv.org/abs/1611.05431) paper for details.
        final_pooling: Optional pooling mode for feature extraction at the final model layer
            when `include_top` is `False`.
            - `None` means that the output of the model
                will be the 4D tensor output of the
                last convolutional layer.
            - `avg` means that global average pooling
                will be applied to the output of the
                last convolutional layer, and thus
                the output of the model will be a
                2D tensor.
            - `max` means that global max pooling will
                be applied.
        top: Defines final layers to evaluate based on a specific problem type. Options are
            'classification' for ImageNet style problems, 'segmentation' for problems like
            the Pascal VOC dataset, and None to exclude these layers entirely.
    Returns:
        The keras `Model`.
    """
```

There are a couple of typos which could be fixed before a pr, sorry about that. The reason it has been changed from true/false to a string parameter is to support both segmentation and classification problems.

Here are the details of papers used for some of the extended features from the top of the file:

```python

"""ResNet v1, v2, and segmentation models for Keras.
# Reference
- [Deep Residual Learning for Image Recognition](https://arxiv.org/abs/1512.03385)
- [Identity Mappings in Deep Residual Networks](https://arxiv.org/abs/1603.05027)
Reference material for extended functionality:
- [ResNeXt](https://arxiv.org/abs/1611.05431) for Tiny ImageNet support.
- [Dilated Residual Networks](https://arxiv.org/pdf/1705.09914) for segmentation support.
- [Deep Residual Learning for Instrument Segmentation in Robotic Surgery](https://arxiv.org/abs/1703.08580)
  for segmentation support.
Implementation Adapted from: github.com/raghakot/keras-resnet

```

**Comment by fchollet** [comment]
@roatienza, @ahundt : any preference for a specific course of action here?

**Comment by roatienza** [comment]
@fchollet - when I was working on Resnet v1 and v2 examples, I discovered that there are different configurations for different datasets. A configuration for ImageNet will have a lower accuracy if applied on CIFAR10/100. 

I made the same mistake on my previous PR for cifar10_resnet.py. I mixed some Resnet config from ImageNet on CIFAR10 example. Hence, this PR fixes that.

While I have not thoroughly investigated, from what I can see the keras contrib example which is derived from https://github.com/raghakot/keras-resnet/blob/master/resnet.py **is ImageNet specific**. This is why they get lower accuracy at 86% compared to >92% from this PR on the simplest ResNet for CIFAR10. The contrib example also has 11.2M parameters for a simple ResNet18 while this PR has only 0.27M parameters for ResNet20. Actually, there is no ResNet18 for CIFAR10 in the original paper. Only ResNet20, 32, 44, 56, 110 and 1202. ResNet18 is one of the configs for ImageNet. 

The same problem exists for the current ResNet50 in the keras/applications. It was optimized for ImageNet not for CIFAR.

In the earlier discussion with @ahundt , I expressed the intention to fix the keras-contrib.

In summary, this PR is a short-term fix for the cifar10_resnet.py. The PR follows the CIFAR10 examples in the paper as close as possible.

Long term, the keras contrib should be moved to applications once it has been fixed and tested to work on both CIFAR and ImageNet datasets. The fixes should have almost the same number of parameters and accuracy as shown in the original Resnet v1 and v2 papers. The future PR for ResNet architecture will replace ResNet50 in the keras/applications. examples/cifar10_resnet.py will then be modified using the ResNet model builder.

**Comment by ahundt** [comment]
> is ImageNet specific. 

is not! :-) I use it on a robot dataset, and I've used it on cifar as well.

File: examples/cifar10_resnet.py [hunk 6]
```diff
@@ -69,87 +100,271 @@
...
+               kernel_size=kernel_size,
+               strides=strides,
+               padding='same',
+               kernel_initializer='he_normal',
+               kernel_regularizer=l2(1e-4))(inputs)
+    x = BatchNormalization()(x)
+    return x
+
+
+def conv_bn_relu(inputs, num_filters=16, kernel_size=3, strides=1):
```
Comment (fchollet) [inline examples/cifar10_resnet.py]: Why not have an `activation` argument in `conv_bn` instead? Seems better than hard-coding `relu` *as a function*. This additional function doesn't seem necessary (same for `bn_relu_conv` -- you could simply have an argument in `conv_bn` to determine where the activation goes).

This is important in particular to make it easy to change the activation easily, and make model easier to parametrize (e.g. if the activation/bn/conv order is a function argument, that's easier to change globally than if it's a function name).

File: examples/cifar10_resnet.py [hunk 7]
```diff
@@ -69,87 +100,271 @@
...
+def resnet_v2(input_shape, depth, num_classes=10):
+    """ResNet Version 2 Model builder [b]
+
+    Stacks of (1 x 1)-(3 x 3)-(1 x 1) BN-ReLU-Conv2D or also known as
+    bottleneck layer
+    First shortcut connection per layer is 1 x 1 Conv2D.
+    Second and onwards shortcut connection is identity.
+    Features maps sizes: 16(input), 64(1st sub_block), 128(2nd), 256(3rd)
+    Torch Reference Code:
+    https://github.com/KaimingHe/resnet-1k-layers/blob/master/resnet-pre-act.lua
```
Comment (fchollet) [inline examples/cifar10_resnet.py]: This is unlicensed code, please don't include a reference to it. Reference to the paper should be enough.

File: examples/cifar10_resnet.py [hunk 8]
```diff
@@ -69,87 +100,271 @@
...
+
+    # Arguments
+        input_shape (tensor): shape of input image tensor
+        depth (int): number of core convolutional layers
+        num_classes (int): number of classes (CIFAR10 has 10)
+
+    # Returns
+        model (Model): Keras model instance
+    """
+    assert (depth - 2) % 9 == 0, 'depth should be 9n+2 (eg 56 or 110 in [b])'
```
Comment (fchollet) [inline examples/cifar10_resnet.py]: Generally, prefer `ValueError` over `assert` as it provides a better error UX.

File: examples/cifar10_resnet.py [hunk 9]
```diff
@@ -166,12 +381,12 @@
     datagen = ImageDataGenerator(
         featurewise_center=False,  # set input mean to 0 over the dataset
         samplewise_center=False,  # set each sample mean to 0
-        featurewise_std_normalization=False,  # divide inputs by std of the dataset
+        featurewise_std_normalization=False,  # divide inputs by std of dataset
         samplewise_std_normalization=False,  # divide each input by its std
         zca_whitening=False,  # apply ZCA whitening
-        rotation_range=0,  # randomly rotate images in the range (degrees, 0 to 180)
-        width_shift_range=0.1,  # randomly shift images horizontally (fraction of total width)
-        height_shift_range=0.1,  # randomly shift images vertically (fraction of total height)
```
Comment (fchollet) [inline examples/cifar10_resnet.py]: These comments are actually useful! If the line is too long, you can put them above each line.

File: examples/cifar10_resnet.py (hunk 6 above)
Comment (roatienza) [reply examples/cifar10_resnet.py]: parameterized resnet_block() function to replace all 3 functions (conv_bn_relu, conv_bn, bn_relu_conv).

File: examples/cifar10_resnet.py (hunk 7 above)
Comment (roatienza) [reply examples/cifar10_resnet.py]: Removed. Thanks.

File: examples/cifar10_resnet.py (hunk 8 above)
Comment (roatienza) [reply examples/cifar10_resnet.py]: Replaced assert by raise ValueError()

File: examples/cifar10_resnet.py (hunk 9 above)
Comment (roatienza) [reply examples/cifar10_resnet.py]: Comments on top of each line.

**Comment by roatienza** [comment]
@fchollet - i pushed a new commit to address the issues raised above. thanks.

**Comment by ozabluda** [comment]
The following is a catalog of differences between the current `cifar10_resnet.py` example and the orig paper. I tried "fixing" some but not all of them, but didn't get statistically significant improvements, doesn't mean it's impossible. Format below is 'orig vs example`

1. Batch size is 128 vs 32. I checked that 128 fits into 12G even with ResNet110. 128 is also much faster than 32 per epoch.
2. NAG with LR=0.1 (with 1-epoch LR=0.01 warmup) vs Adam with LR=0.001
3. No LR reduction on plateau for CIFAR (only for ImageNet) vs broken ReduceLROnPlateau.
4. 154 epochs (60k iterations) vs 200 epochs.
5. LR 10x reduction schedule after 32k iterations (82 epochs), 48k iterations (123 epochs) stop at 60k iterations (154 epochs). Vs 80,120,160,180. Note that in the orig paper epoch=45k samples, the previous sentence recalculated for epoch=50k samples in the example.
6. Subtractive pixelwise (mean-image) normalization vs also divisive normalization (/255).
7. Weight decay of 1e-4 vs l2=1e-4 (which is 2x greater).
8. 4 pixels translation jitter vs 3.2 pixels.
9. `fill_mode='constant'` (black) vs `fill_mode='nearest'`
10. Probably bilinear interpolation vs nearest.
 
Should I document these in the `cifar10_resnet.py` itself?

**Comment by roatienza** [comment]
@ozabluda : This is what I did in my experiments:
1). I built the example on 1080Ti (11G) so I have to use smaller batch size. I did not push the batch size to 64. It might work there as well.
2) I recall using the same optimizer as the one in paper but got better results using Adam.
3) Got better results using  ReduceLROnPlateau vs no LR reduction
4) I think I stretched it to 200epochs just to see any further reduction. I recall that are still small improvements beyond 154.
5) The schedule is a close approximation of the paper's. 
6 to 10) I used the closest approximation of experimental setup based on the paper. Pls make a PR if you get statistically better results.

**Comment by ozabluda** [comment]
>@roatienza 
3) Got better results using ReduceLROnPlateau vs no LR reduction

This was a statistical fluctuation, because  `LearningRateScheduler` always overrides `ReduceLROnPlateau`, see PR https://github.com/keras-team/keras/pull/8829

**Comment by roatienza** [comment]
@ozabluda 

I think that was the intention there. Originally, it was only `ReduceLROnPlateau` but it appears that sometimes it is too late in reducing the lr at higher epoch. Late lr reduction prevents resnet from achieving higher validation acc. When it is late, `LearningRateScheduler` kicks in. If however, `LearningRateScheduler` always overrides `ReduceLROnPlateau` then the latter has no effect and should be removed.

**Comment by ozabluda** [comment]
Soliciting comments for logic and docs improvements, particularly naming https://github.com/keras-team/keras/pull/8955#issuecomment-355350523
